📖 Add APIBindings example

SimonTheLeg opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary

Adds an APIBinding example to fit with the example previously on the page and adds some more information on APIBindings/turns some of the bullets into full text.

The asciiflow link did not resolve for me and is also not rendered in the documentation at the moment, so I removed it.

## Related issue(s)

Fixes #

## Release Notes

<!--
Please add a release note in the block below. Leave NONE only if no user-facing changes are in this PR.
-->

```release-note
NONE
```

sttts commented:
/ok-to-test

sttts commented on docs/content/concepts/apis/exporting-apis.md:
```diff
@@ -374,12 +374,33 @@ TODO
+
+`APIBindings` are used to import API resources. They contain a reference to an `APIExport` using the namespace and workspace path of an `APIExport` and will bind all APIs defined in the `APIExport`. The reference path needs to be provided to you by the provider of the API or an external catalog solution.
+
+Furthermore, `APIBindings` provide access to the groups/resources defined in an `APIExport's` `PermissionClaims` list.
```
```suggestion
Furthermore, `APIBindings` provide the `APIExport` owner access to additional resources defined in an `APIExport`'s `PermissionClaims` list. `PermissionClaims` must be accepted by the user explicitly, before this access is granted. The resources can be builtin Kubernetes resources or resources from other `APIExports`.
```

sttts replied:
Maybe also:

```
When an `APIExport` is changed after workspaces bound to them, new or changed APIs are automatically propagated to all `APIBindings`. New `PermissionClaims` on the other hand are NOT automatically accepted.

SimonTheLeg replied:
this is a really good point! Thanks I have added it

sttts commented on docs/content/concepts/apis/exporting-apis.md:
````diff
@@ -374,12 +374,33 @@ TODO
+      path: "root:api-provider" # path of your api-provider workspace
+```
+
+It should be noted that `APIBindings` do not create `CRDs` or `APIResourceSchemas`. Instead APIs are directly bound. This means you can query for imported APIs using `kubectl api-resources`. Additionally you can use `kubectl explain` to get a detailed view on all fields of the API.
````
```suggestion
It should be noted that `APIBindings` do not create `CRDs` or `APIResourceSchemas`. Instead APIs are directly bound. 

Bound APIs are like any other resources in kcp or Kubernetes. This means you can query for imported APIs using `kubectl api-resources`. Additionally you can use `kubectl explain` to get a detailed view on all fields of the API.
```

embik commented on docs/content/concepts/apis/exporting-apis.md:
```diff
@@ -374,12 +374,36 @@ TODO
-
-[diagram1]: https://asciiflow.com/#/share/eJyrVspLzE1VssorzcnRUcpJrEwtUrJSqo5RqohRsrI0NdGJUaoEsozMzYCsktSKEiAnRkmBGPBoyh5qoZiYPGKtVFBwzs8rLs1NLVIIzy%2FKLi5ITE6FyJBgyIC4G5cMEYZgtVwhPDMlPbWkWMExwNMpMy8lMy%2BdFAOp5C44BXGNgiMWY6gY4igBgNUBTtgdAGQDw0khoCi%2FLDMFNfHgNMp5gPxCxeSJO4YR8YeqEilVuVYU5BeVKDya3kKCDdj5ONROw68WyS1BqcX5pUXJqcHJGam5iehx1vNoSgM10AT6xHATzlKsiZRcN4dKvl5C1xIDS9DgKMmICQyoqU24ZUgyBEcpRpY ...
+
+`APIBindings` are used to import API resources. They contain a reference to an `APIExport` using the namespace and workspace path of an `APIExport` and will bind all APIs defined in the `APIExport`. The reference path needs to be provided to you by the provider of the API or an external catalog solution.
```
```suggestion
`APIBindings` are used to import API resources. They contain a reference to an `APIExport` using the name and workspace path of an `APIExport` and will bind all APIs defined in the `APIExport`. The reference path needs to be provided to you by the provider of the API or an external catalog solution if you don't have direct access to the workspace containing the `APIExport`.
```

I think the "namespace" part here is wrong, APIExports are cluster-level resources. Maybe "name" was meant? In addition, I suggest a little clarification on how to get the APIExport path (if you have access directly, you don't need someone to give you the path; it's entirely possible that people make this choice).

mjudeikis replied:
`using the name of the APIExport and path to the workspace, where an APIExport is located` 
something like that maybe

SimonTheLeg replied:
fixed

embik commented on docs/content/concepts/apis/exporting-apis.md:
```diff
@@ -374,12 +374,36 @@ TODO
+
+`APIBindings` are used to import API resources. They contain a reference to an `APIExport` using the namespace and workspace path of an `APIExport` and will bind all APIs defined in the `APIExport`. The reference path needs to be provided to you by the provider of the API or an external catalog solution.
+
+Furthermore, `APIBindings` provide the `APIExport` owner access to additional resources defined in an `APIExport`'s `PermissionClaims` list. `PermissionClaims` must be accepted by the user explicitly, before this access is granted. The resources can be builtin Kubernetes resources or resources from other `APIExports`.
```
Nit: I think we shouldn't put permission claims in backticks, they aren't a standalone resource type. It would be more useful to put the JSONpath references to where permission claims are configured in `APIExport` and `APIBindings`.

SimonTheLeg replied:
fixed

embik commented on docs/content/concepts/apis/exporting-apis.md:
````diff
@@ -374,12 +374,36 @@ TODO
+      path: "root:api-provider" # path of your api-provider workspace
+```
+
+It should be noted that `APIBindings` do not create `CRDs` or `APIResourceSchemas`. Instead APIs are directly bound. 
````
```suggestion
It should be noted that `APIBindings` do not create `CRDs` or `APIResourceSchemas` in the workspace. Instead APIs are directly bound. 
```

Strictly speaking (as far as I know), CRDs are created somewhere in the shadows, so let's make this sentence a little more precise.

embik commented on docs/content/concepts/apis/exporting-apis.md:
````diff
@@ -374,12 +374,36 @@ TODO
+widgets                example.kcp.io/v1alpha1   false        Widget
+```
+
+Furthermore, you can use the `APIBinding.Status.BoundResources` field to precisely identify which `APIResourceSchemas` have been imported.
````
```suggestion
Furthermore, you can use the `.status.boundResources` field of an `APIBinding` to precisely identify which `APIResourceSchemas` have been imported.
```

I would say we should reference via JSONpath, not Go structs. That only makes sense if you develop a controller, and you usually don't write controllers that check an APIBinding status.

SimonTheLeg replied:
yes good point!

embik reviewed (changes requested):
Some small changes I would suggest.

janwillies commented on docs/content/concepts/apis/exporting-apis.md:
````diff
@@ -374,12 +374,36 @@ TODO
+      path: "root:api-provider" # path of your api-provider workspace
+```
+
+It should be noted that `APIBindings` do not create `CRDs` or `APIResourceSchemas`. Instead APIs are directly bound. 
````
> Instead APIs are directly bound.
What does this mean?

mjudeikis replied:
maybe something "apis are reflected". It there, but not in the same way in Kubernetes. So some clearing the waters is needed

SimonTheLeg replied:
I have re-worked this section again. I would really like to keep the word "bound" in as it comes directly from how Kubernetes behind the scenes calls it "binding apis".

I have made it more clear that this is something Kubernetes does for you behind the scenes and then in the next sentence I outlined what the practical implications of it are.
I think it is important in this section that we don't go into too many details and only outline what this means for an end-user.

jmcshane commented on docs/content/concepts/apis/exporting-apis.md:
```diff
@@ -374,12 +374,36 @@ TODO
+
+`APIBindings` are used to import API resources. They contain a reference to an `APIExport` using the namespace and workspace path of an `APIExport` and will bind all APIs defined in the `APIExport`. The reference path needs to be provided to you by the provider of the API or an external catalog solution.
+
+Furthermore, `APIBindings` provide the `APIExport` owner access to additional resources defined in an `APIExport`'s `PermissionClaims` list. `PermissionClaims` must be accepted by the user explicitly, before this access is granted. The resources can be builtin Kubernetes resources or resources from other `APIExports`.
```
How are these permissions granted? Need an example in line with the `APIExport` sample above.

SimonTheLeg replied:
very good point. I have added an example which matches the previous APIExport

SimonTheLeg commented:
@embik I think this PR is ready for review/merging now. I have addressed all the comments and I think it's a good state to merge. 
More improvements can be added in additional PRs

embik reviewed (changes requested):
One last (🤞🏻 ) thing.

embik reviewed (approved):
/approve

SimonTheLeg commented:
/cherry-pick release-0.27
